### PR TITLE
Abort connection if no QUIC transport parameters are found

### PR DESCRIPTION
**Pull request by djc**
cc @Ralith

### CODE REVIEW

File: rustls/src/conn.rs [hunk 1]
```diff
@@ -958,6 +958,12 @@ impl ConnectionCommon {
             .prepare_message_decrypter(dec);
     }
 
+    #[cfg(feature = "quic")]
+    pub fn missing_extension(&mut self, why: &str) -> Error {
+        self.send_fatal_alert(AlertDescription::MissingExtension);
+        Error::PeerIncompatibleError(why.to_string())
```
Comment (briansmith) [inline rustls/src/conn.rs:964]: Seems like peer misbehavior, since the extensions are required for QUIC to work, right?

Comment (djc) [reply rustls/src/conn.rs]: Yup, copy/pasta error.

File: rustls/src/quic.rs [hunk 2]
```diff
@@ -32,6 +32,12 @@ impl Secrets {
 /// Generic methods for QUIC sessions
 pub trait QuicExt {
     /// Return the TLS-encoded transport parameters for the session's peer.
+    ///
+    /// While the transport parameters are technically available prior to the
+    /// completion of the handshake, they cannot be fully trusted until the
+    /// handshake completes, and reliance on them should be minimized.
+    /// However, any tampering with the parameters will cause the handshake
+    /// to fail.
     fn get_quic_transport_parameters(&self) -> Option<&[u8]>;
```
Comment (briansmith) [inline rustls/src/quic.rs:41]: We should file a separate issue for this: We should refactor this API to minimize the danger here.

Comment (djc) [reply rustls/src/quic.rs:41]: Yeah, I wanted to get @Ralith's eyes on it before changing it. While reviewing code in Quinn, it did seem like it relies on being able to access this early. (The change I had in mind was just to return `None` while `self.is_handshaking()`.)

**Review by briansmith — APPROVED** [review]
Not sure what our QUIC testing situation is. Seems like we should improve that so we can write tests for things like this.

**Comment by Ralith** [comment]
See `test_quic` in rustls/tests/api.rs for current testing. We indeed do not have coverage for this case.

File: rustls/src/quic.rs (hunk 2 above)
Comment (Ralith) [reply rustls/src/quic.rs:41]: Responded in #655.

**Review by Ralith — APPROVED** [review]
This is consistent with the QUIC spec. We're currently handling this case explicitly in Quinn, but I think moving it into rustls is a bit nicer because this way quinn's ostensibly crypto-agnostic code doesn't need a hardcoded TLS alert number.

**Comment by djc** [comment]
Going to close this in favor of a new PR that adds tests and address some other issues.

**Comment by djc** [comment]
See #666.

**Pull request closed without merging**